### PR TITLE
feat(issue-249): add support to mock k8s api

### LINKED ISSUE #249: How to test Lookup function Scenario with helm unittest
I've stumbled upon this [closed issue - 108](https://github.com/helm-unittest/helm-unittest/issues/108) asked in year 2020 which refer to use older code base. Did we introduced the same functionality in this repo?  If so, can someone point me towards related documentation.

### PR DESCRIPTION
**Pull request by cgroschupp**
Fixes #249

### CODE REVIEW

**Comment by cgroschupp** [comment]
@quintush Any suggestions for improvement?

File: test/data/v3/with-k8s-fake-client/templates/lookup.yaml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+---
+lookup_exists: {{ (lookup "v1" "Pod" "default" "exists") | toYaml | nindent 2 }}
+lookup_not_exists: {{ (lookup "v1" "Pod" "default" "not-exists") | toYaml | nindent 2 }}
```
Comment (ivankatliarchuk) [inline test/data/v3/with-k8s-fake-client/templates/lookup.yaml:3]: new line is missing

Comment (ivankatliarchuk) [reply test/data/v3/with-k8s-fake-client/templates/lookup.yaml:3]: Could this help with testing
```
{{- if .Capabilities.APIVersions.Has "policy/v1/PodDisruptionBudget" }}
apiVersion: policy/v1
{{- else }}
apiVersion: policy/v1beta1
{{- end }}
```

If yes, how example test is going to looks like?

Comment (cgroschupp) [reply test/data/v3/with-k8s-fake-client/templates/lookup.yaml:3]: No, this feature is only for the lookup function.

But you can already write tests to test against the apiVersion.
```
capabilities:
  apiVersions:
    - policy/v1
```

**Comment by cgroschupp** [comment]
@ivankatliarchuk i have updated my code, please check again.

File: test/data/v3/with-k8s-fake-client/tests/lookup_test.yaml [hunk 2]
```diff
@@ -0,0 +1,20 @@
+templates:
```
Comment (ivankatliarchuk) [inline test/data/v3/with-k8s-fake-client/tests/lookup_test.yaml:1]: possibly more example needed.

`kubernetes_provider` in top level is missing as well

File: pkg/unittest/test_job.go [hunk 3]
```diff
@@ -23,6 +29,36 @@ import (
...
+	Gvr        schema.GroupVersionResource `yaml:"gvr"`
+	Namespaced bool                        `yaml:"namespaced"`
+}
+
+type KubernetesFakeClientProvider struct {
+	Scheme  map[string]KubernetesFakeKindProps `yaml:"scheme"`
+	Objects []map[string]interface{}           `yaml:"objects"`
+}
+
+func (p *KubernetesFakeClientProvider) GetClientFor(apiVersion, kind string) (dynamic.NamespaceableResourceInterface, bool, error) {
```
[the remark this answers is not in the document]
Comment (cgroschupp) [reply pkg/unittest/test_job.go]: added tests

File: pkg/unittest/test_job.go [hunk 4]
```diff
@@ -23,6 +28,36 @@ import (
 	v3engine "helm.sh/helm/v3/pkg/engine"
 )
 
+type KubernetesFakeKindProps struct {
```
[the remark this answers is not in the document]
Comment (cgroschupp) [reply pkg/unittest/test_job.go]: i decided to use the name `fake`.

File: test/data/v3/with-k8s-fake-client/tests/lookup_test.yaml (hunk 2 above)
Comment (cgroschupp) [reply test/data/v3/with-k8s-fake-client/tests/lookup_test.yaml:1]: i added more examples.

What do you mean kubernetes_provider is missing at the top level?

File: test/data/v3/with-k8s-fake-client/templates/lookup.yaml [hunk 5]
```diff
@@ -0,0 +1,8 @@
+---
+pod_exists: {{ (lookup "v1" "Pod" "default" "exists") | toYaml | nindent 2 }}
+pod_not_exists: {{ (lookup "v1" "Pod" "default" "not-exists") | toYaml | nindent 2 }}
+ingress_exists: {{ (lookup "networking.k8s.io/v1" "Ingress" "default" "exists") | toYaml | nindent 2 }}
+ingress_not_exists: {{ (lookup "networking.k8s.io/v1" "Ingress" "default" "not-exists") | toYaml | nindent 2 }}
+namespace_exists: {{ (lookup "v1" "Namespace" "" "exists") | toYaml | nindent 2 }}
+namespace_not_exists: {{ (lookup "v1" "Namespace" "" "not-exists") | toYaml | nindent 2 }}
+namespaces: {{ (lookup "v1" "Namespace" "" "") | toYaml | nindent 2 }}
```
Comment (ivankatliarchuk) [inline test/data/v3/with-k8s-fake-client/templates/lookup.yaml:8]: new line missing

File: test/data/v3/with-k8s-fake-client/tests/lookup_test.yaml [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+            name: exists
+            namespace: default
+        - kind: Namespace
+          apiVersion: v1
+          metadata:
+            name: exists
+        - kind: Namespace
+          apiVersion: v1
+          metadata:
+            name: exists2
```
Comment (ivankatliarchuk) [inline test/data/v3/with-k8s-fake-client/tests/lookup_test.yaml:70]: new line missing

File: pkg/unittest/k8s_fake_client_test.go [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+		"apiVersion": apiVersion,
+		"kind":       kind,
+		"metadata": map[string]interface{}{
+			"namespace": namespace,
+			"name":      name,
+		},
+	}
+}
+
+func TestKubernetesFakeClient(t *testing.T) {
```
Comment (ivankatliarchuk) [inline pkg/unittest/k8s_fake_client_test.go:23]: Probably worth to review naming.

```
KubernetesProvider kubernetesFakeClientProvider `yaml:"kubernetesProvider"`
```

and test tfunction name

```
func TestKubernetesFakeClient
```

**Review by ivankatliarchuk — COMMENTED** [review]
added review. rest lgtm.

~~It's worth to synchronize with the master branch so that the CI can run tests against these changes.~~

**Comment by ivankatliarchuk** [comment]
I'm not a mantainer btw. From high level interesting capability.

Just consider naming. There are requets to mock `helm` functions too. So the naming could be

```
kubernetesProvider
helmProvider
```

or 

```
kubernetesMock
helmMock
```

and etc.

File: test/data/v3/with-k8s-fake-client/tests/lookup_test.yaml (hunk 2 above)
Comment (ivankatliarchuk) [reply test/data/v3/with-k8s-fake-client/tests/lookup_test.yaml:1]: example. Top level config e.g. suite level.

```yml
templates:
  - templates/lookup.yaml
 kubernetesProvider:
      scheme:
        "v1/Namespace":
          gvr:
            version: "v1"
            resource: "namespaces"
          namespaced: false
        "v1/Pod":
          gvr:
            version: "v1"
            resource: "pods"
          namespaced: true
        "networking.k8s.io/v1/Ingress":
          gvr:
            group: "networking.k8s.io"
            version: "v1"
            resource: "ingresses"
          namespaced: true
tests:
  - it: manifest should match snapshot
    asserts:     
      - equal:
          path: pod_exists.metadata.name
          value: exists
      - equal:
          path: pod_exists.metadata.namespace
          value: default
      - lengthEqual:
          path: namespaces.items
          count: 2
      objects:
        - kind: Pod
          apiVersion: v1
          metadata:
            name: exists
            namespace: default
```

Comment (ivankatliarchuk) [reply test/data/v3/with-k8s-fake-client/tests/lookup_test.yaml:1]: probably worth to consider something like ^ for a future release

**Comment by cgroschupp** [comment]
@quintush can you please review this?

**Pull request merged**